js/ui/grid_core/ui.grid_core.column_chooser.js

```diff
@@ -369,38 +370,38 @@ const columnChooserMembers = {
         this._columnChooserList.option('items', items);
     },
 
-    _updateItemSelection: function(columnIndex) {
-        const isRecursive = this.option('columnChooser.selection.recursive');
-        const column = this._columnsController.columnOption(columnIndex);
+    _updateItemsSelection: function(columnIndices) {
+        const columnChooserList = this._columnChooserList;
+        const changedColumns = columnIndices?.map((columnIndex) => this._columnsController.columnOption(columnIndex));
 
-        const isRecursiveWithColumns = isRecursive && column.hasColumns;
-
-        if(!isRecursiveWithColumns) {
-            column.visible
-                ? this._columnChooserList.selectItem(columnIndex)
-                : this._columnChooserList.unselectItem(columnIndex);
-        }
+        columnChooserList.beginUpdate();
+        changedColumns?.forEach((changedColumn) => {
+            if(changedColumn.visible) {
+                columnChooserList.selectItem(changedColumn.index);
+            } else {
+                columnChooserList.unselectItem(changedColumn.index);
+            }
+        });
+        columnChooserList.endUpdate();
     },
 
     _columnOptionChanged: function(e) {
         this.callBase(e);
 
-        const changeTypes = e.changeTypes;
-        const optionNames = e.optionNames;
         const isSelectMode = this.isSelectMode();
 
         if(isSelectMode && this._columnChooserList && this._isUpdatingColumnVisibility !== true) {
-            const onlyOneColumnChanged = e.columnIndex !== undefined;
+            const optionNames = e.optionNames;
```

Comment:
Minor comment.

Let's use destructuring here:
```suggestion
            const { optionNames } = e;
```